### ISSUE TITLE
mdadm is unconditionally installed, even when the boot.initrd.services.swraid.enable is set to false

### ISSUE THREAD

**Issue by filakhtov** [NixOS/nixpkgs#215642]
### Describe the bug
👋🏻hello lovely people!

I am at the beginning of my journey playing with Nix/NixOS and I am planning to use NixOS to build images on top of it for my LXC (and potentially later Docker) images. At the moment I use the `nixos-generate -f lxc` to produce LXC container images. However, the resulting images come pretty chunky and I am working on trimming a lot of unneeded stuff down and this is where I noticed that the `mdadm` package is unconditionally installed into the system, even though in case of LXC this would make little to no sense. Being a newbie, it took me a good while to figure out what nixpkgs modules are and identify the source of the problem, the [`swraid.nix`](https://github.com/NixOS/nixpkgs/blob/af96094e9b8eb162d70a84fa3b39f4b7a8b264d2/nixos/modules/tasks/swraid.nix#L20) module. As can be seen, the `mdadm` package is always added to the `environment.systemPackages` outside of the `mkIf` check for the `enable` option. I believe this is incorrect and should be fixed. Please let me know if I am missing something and thanks in advance!

### Steps To Reproduce
1. Install `nixos-generate` or just run it in the `nix-shell`;
2. Run `nixos-generate -f lxc` to produce a base NixOS LXC image, note the resulting path output;
3. Run the container using the produced LXC image;
4. Check `mdadm` presence via any means you like (`nix-store --query --requisites /run/current-system`, `nix repl`, etc);
5. Set `boot.initrd.services.swraid.enable = lib.mkForce false` option in the `configuration.nix` for the container;
6. Run `nixos-rebuild switch`;
7. Check `mdadm` presence via any means you like - it is still there;

### Expected behavior
`mdadm` to be removed from the `environment.systemPackages` if the `boot.initrd.services.swraid.enable` set to `false`

### Screenshots
Not applicable

### Additional context
This problem doesn't have to be reproduced through LXC containers either. Any NixOS system would have `mdadm` installed, regardless if the underlying system is using the Raid configuration or not.

### Notify maintainers
The `swraid` doesn't have any maintainers listed, so I am unsure who to ping, but @dasJ was the last to actually do something significant in it, so that's who I'll ping.

### Metadata
```console
[root@nixos:~]# nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 5.15.83-1-pve, NixOS, 23.05 (Stoat), 23.05pre451903.5a350a8f31b`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.13.2`
 - channels(root): `"nixos"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```

**Comment by filakhtov** [comment]
Just to follow up and hopefully get this issue moving - is my thinking correct here that the fix would be as simple as adding the `lib.mkIf` check, like so:


```nix
{
    # ... snip ...
    config = {
        environment.systemPackages = lib.mkIf cfg.enable [ pkgs.mdadm ];

        services.udev.packages = lib.mkIf cfg.enable [ pkgs.mdadm ];

        systemd.packages = lib.mkIf cfg.enable [ pkgs.mdadm ];

        # ... snip ...
}
```


If that is the case I am happy to open up a pull request if anyone would be able and willing to review and help merge it.

**Comment by Majiir** [comment]
A system may still be using RAID without `boot.initrd.services.swraid.enable`. Would removing these lines break those systems?

**Comment by dasJ** [comment]
> boot.initrd.services.swraid.enable = lib.mkForce false

This option is only relevant for the new initr (which is not done iirc). I didnt want to touch more than the initrd because I was afraid of regressions.

**Comment by filakhtov** [comment]
@Majiir systemd does not depend on `mdadm` to my knowledge. There are no library dependencies, nor any unit files that reference the binary. `mdadm` provides it's own unit files when installed. So I believe it should be safe to nuke. I am happy to open up a PR and perform whatever testing is necessary.

**Comment by filakhtov** [comment]
I have been looking into this a bit more and testing a few things and I think there are two things that can be practically done in order to avoid breaking changes:

- either we make `boot.initrd.services.swraid.enable` set to true by default and add the `mkIf` as I proposed above;
- or we create a new flag `boot.initrd.services.swraid.disable` which can conflict with `enable` and get rid of these packages from the system package set.

What do you think @dasJ @Majiir?

**Comment by stelcodes** [comment]
This was fixed by #183314, closing now.
